### PR TITLE
app secure authorized by app_user_id

### PR DESCRIPTION
**Pull request by NantasakThiti**

### CODE REVIEW

File: Cpp/odin-views/app_secure.cpp [hunk 1]
```diff
@@ -88,9 +99,50 @@ namespace {
                     if (jwt.value().payload.has_key("sub")) {
                         req.headers().set("__jwt", jwt.value().payload, "sub");
                         req.headers().set(
-                                "__user",
+                                "__app_user",
                                 fostlib::coerce<fostlib::string>(
                                         jwt.value().payload["sub"]));
+                        // select
+                        static const fostlib::string sql(
+                                "SELECT identity_id FROM "
```
Comment (tleekkul) [inline Cpp/odin-views/app_secure.cpp]: east const
and move to outside of method

Comment (tleekkul) [reply Cpp/odin-views/app_secure.cpp]: edit naming

File: Cpp/odin-views/app_secure.cpp [hunk 2]
```diff
@@ -88,9 +99,50 @@ namespace {
...
+
+                        auto app =  fostlib::coerce<fostlib::string>(
+                                    jwt.value().payload["iss"])
+                                    .substr(odin::c_app_namespace.value()
+                                    .code_points());
+
+                        auto user = fostlib::coerce<fostlib::string>(
+                                    jwt.value().payload["sub"]);
+
+                        auto data = fostgres::sql(
```
Comment (tleekkul) [inline Cpp/odin-views/app_secure.cpp]: auto userid

File: Cpp/odin-views/app_secure.cpp [hunk 3]
```diff
@@ -88,9 +99,50 @@ namespace {
...
+                                "SELECT identity_id FROM "
+                                "odin.app_user "
+                                "WHERE app_id=$1 AND app_user_id=$2");   
+
+                        auto app =  fostlib::coerce<fostlib::string>(
+                                    jwt.value().payload["iss"])
+                                    .substr(odin::c_app_namespace.value()
+                                    .code_points());
+
+                        auto user = fostlib::coerce<fostlib::string>(
```
Comment (tleekkul) [inline Cpp/odin-views/app_secure.cpp]: auto app_user_id

File: Cpp/odin-views/app_secure.cpp [hunk 4]
```diff
@@ -88,9 +99,50 @@ namespace {
...
+                                "__app_user",
                                 fostlib::coerce<fostlib::string>(
                                         jwt.value().payload["sub"]));
+                        // select
+                        static const fostlib::string sql(
+                                "SELECT identity_id FROM "
+                                "odin.app_user "
+                                "WHERE app_id=$1 AND app_user_id=$2");   
+
+                        auto app =  fostlib::coerce<fostlib::string>(
```
Comment (tleekkul) [inline Cpp/odin-views/app_secure.cpp]: const

File: Cpp/odin-views/app_secure.cpp [hunk 5]
```diff
@@ -88,9 +99,50 @@ namespace {
...
+                                    jwt.value().payload["iss"])
+                                    .substr(odin::c_app_namespace.value()
+                                    .code_points());
+
```
Comment (tleekkul) [inline Cpp/odin-views/app_secure.cpp]: # check existing of "iss", "sub" of jwt payload are in the load_jwt()

File: Cpp/odin-views/app_secure.cpp [hunk 6]
```diff
@@ -13,12 +13,23 @@
...
+    inline std::pair<boost::shared_ptr<fostlib::mime>, int>
+            respond(fostlib::string message, int code = 403) {
+        fostlib::json ret;
+        if (not message.empty())
+            fostlib::insert(ret, "message", std::move(message));
+        fostlib::mime::mime_headers headers;
+        boost::shared_ptr<fostlib::mime> response(new fostlib::text_body(
+                fostlib::json::unparse(ret, true), headers, "application/json"));
+        return std::make_pair(response, code);
+    }
```
Comment (tleekkul) [inline Cpp/odin-views/app_secure.cpp]: On Line#43: `jwt_iss.find(odin::c_app_namespace.value())` -> Should use `starts_with` instead of `find`

File: Cpp/odin-views/app_secure.cpp (hunk 2 above)
Comment (tleekkul) [reply Cpp/odin-views/app_secure.cpp]: can use cnx.select

File: Cpp/odin-views/app_secure.cpp [hunk 7]
```diff
@@ -88,9 +99,50 @@ namespace {
...
+                        auto data = fostgres::sql(
+                                cnx, sql,
+                                std::vector<fostlib::string>{app, user});  
+
+                        auto &rs = data.second;
+                        
+                        auto row = rs.begin();
+                        if (row == rs.end()) {
+                          return respond("App user does not exists.", 401);
+                        }
```
Comment (tleekkul) [inline Cpp/odin-views/app_secure.cpp]: exist(s) no 's'

File: Cpp/odin-views/app_secure.cpp [hunk 8]
```diff
@@ -88,9 +99,50 @@ namespace {
...
+                                }
+                                fostlib::jcursor pos;
+                                for (const auto &p : parts) pos /= p;
+                                fostlib::insert(app_json, pos, record[index]);
+                        }
+                        req.headers().set(
+                                "__user",
+                                fostlib::coerce<fostlib::string>(
+                                    app_json["identity_id"]
+                                ));
```
Comment (tleekkul) [inline Cpp/odin-views/app_secure.cpp]: just try record[0]

File: Cpp/odin-fg/native.jwt.cpp [hunk 9]
```diff
@@ -69,8 +69,12 @@ const fg::frame::builtin odin::lib::jwt_payload =
         [](fg::frame &stack,
            fg::json::const_iterator pos,
            fg::json::const_iterator end) {
+            fostlib::string app_id= "";
+            if (pos!=end) {
+                app_id = stack.resolve_string(stack.argument("app_id", pos, end));
```
Comment (KayEss) [inline Cpp/odin-fg/native.jwt.cpp]: I don't think the app_id should be the first argument. I think the JWT itself should be first, and then the app ID could follow.

File: Schema/authn/004-not-allow-merge-registerd-to-guest.blue.sql [hunk 10]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO odin.migration VALUES('authn', '004-not-allow-merge-registerd-to-guest.blue.sql');
```
Comment (KayEss) [inline Schema/authn/004-not-allow-merge-registerd-to-guest.blue.sql]: Spelled "registered"

File: Cpp/odin-fg/native.jwt.cpp [hunk 11]
```diff
@@ -69,17 +69,20 @@ const fg::frame::builtin odin::lib::jwt_payload =
         [](fg::frame &stack,
            fg::json::const_iterator pos,
            fg::json::const_iterator end) {
-            auto jwt = fostlib::jwt::token::load(
-                    odin::c_jwt_secret.value(),
-                    fostlib::coerce<fostlib::string>(stack.lookup("odin.jwt")));
-            if (not jwt) {
+            auto jwt = stack.resolve_string(stack.argument("jwt", pos, end));
+            auto secret = odin::c_jwt_secret.value();
```
Comment (tleekkul) [inline Cpp/odin-fg/native.jwt.cpp]: jwt and app_id must be optional

File: Cpp/odin-views/app_user.cpp [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+#include <fost/json>
+#include <fost/log>
+#include <fost/push_back>
+#include <fostgres/sql.hpp>
+
+namespace {
+
+    fostlib::json get_app_user(
+            fostlib::pg::connection &cnx, const f5::u8view app_id, const f5::u8view app_user_id) {
+        static const f5::u8view sql(
```
Comment (tleekkul) [inline Cpp/odin-views/app_user.cpp]: delete static and {} for constructor

File: Cpp/odin-views/app_user.cpp [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        auto &rs = data.second;
+        auto row = rs.begin();
+        if (row == rs.end()) {
+            fostlib::log::warning(odin::c_odin)("", "App not found")("app_id", app_id)("app_user_id", app_user_id);
+            return fostlib::json();
+        }
+        auto record = *row;
+        fostlib::json app;
+        for (std::size_t index{0}; index < record.size(); ++index) {
+            const auto parts = fostlib::split(data.first[index], "__");
```
Comment (tleekkul) [inline Cpp/odin-views/app_user.cpp]: we dont't want __

File: Cpp/odin-views/app_user.cpp [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+                fostlib::json config;
+                fostlib::insert(config, "view", "fost.response.405");
+                fostlib::push_back(config, "configuration", "allow", "PUT");
+                return execute(config, path, req, host);
+            }
+
+            auto parameters = fostlib::split(path, '/');
+            if (parameters.size() != 2) {
+                boost::shared_ptr<fostlib::mime> response(
+                    new fostlib::text_body(fostlib::string("400 Bad Request\n"),
```
Comment (tleekkul) [inline Cpp/odin-views/app_user.cpp]: should be 404

File: Cpp/odin-views/app_user.cpp [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+                        L"text/plain"));
+                return std::make_pair(response, 400);
+            }
+            auto const app_id = parameters[0];
+            auto const app_user_id = parameters[1];
+            auto const body_str = fostlib::coerce<fostlib::string>(
+                    fostlib::coerce<fostlib::utf8_string>(req.data()->data()));
+            auto const app_data = fostlib::json::parse(body_str);
+            fostlib::pg::connection cnx{fostgres::connection(config, req)};
+            auto app_user = get_app_user(cnx, parameters[0], parameters[1]);
```
Comment (tleekkul) [inline Cpp/odin-views/app_user.cpp]: app_id and app_user_id

File: Cpp/odin-views/app_user.cpp [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+                boost::shared_ptr<fostlib::mime> response(
+                    new fostlib::text_body(fostlib::string("400 Bad Request\n"),
+                        fostlib::mime::mime_headers(),
+                        L"text/plain"));
+                return std::make_pair(response, 400);
+            }
+            auto const app_id = parameters[0];
+            auto const app_user_id = parameters[1];
+            auto const body_str = fostlib::coerce<fostlib::string>(
+                    fostlib::coerce<fostlib::utf8_string>(req.data()->data()));
```
Comment (tleekkul) [inline Cpp/odin-views/app_user.cpp]: don't need coercion if using req.data()->body_as_string()

File: Cpp/odin-views/app_user.cpp [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+            auto const app_data = fostlib::json::parse(body_str);
+            fostlib::pg::connection cnx{fostgres::connection(config, req)};
+            auto app_user = get_app_user(cnx, parameters[0], parameters[1]);
+            if (app_user.isnull()) {
+                auto identity_id = odin::reference();
+                auto const reference = odin::reference();
```
Comment (tleekkul) [inline Cpp/odin-views/app_user.cpp]: should use reference from higher view if it exists.

File: Cpp/odin-views/app_user.cpp [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+                fostlib::insert(new_app_user, "reference", reference);
+                cnx.insert("odin.app_user_ledger", new_app_user);
+
+                fostlib::json new_app_data;
+                fostlib::insert(new_app_data, "app_id", app_id);
+                fostlib::insert(new_app_data, "identity_id", identity_id);
+                fostlib::insert(new_app_data, "app_data", app_data);
+                fostlib::insert(new_app_data, "reference", reference);
+                cnx.insert("odin.app_user_app_data_ledger", new_app_data);
+                cnx.commit();
```
Comment (tleekkul) [inline Cpp/odin-views/app_user.cpp]: move commit out of scope.

File: Cpp/odin/credentials.cpp [hunk 19]
```diff
@@ -15,6 +15,81 @@
...
 
+fostlib::json odin::app_credentials(
+        fostlib::pg::connection &cnx,
+        const fostlib::string &username,
+        const fostlib::string &password,
+        const fostlib::string &app_id,
+        const fostlib::nullable<fostlib::host> &source,
+        const fostlib::json &annotation) {
+    static const fostlib::string sql(
+            "SELECT "
```
Comment (tleekkul) [inline Cpp/odin/credentials.cpp]: must be f5::u8view

File: tests/authnz.fg [hunk 20]
```diff
@@ -27,7 +28,7 @@ set-path testserver.headers ["Authorization"] ""
 GET odin/api / 401
 
 # Check that the default JWT contains the subject key
-set payload (odin.jwt.payload)
+set payload (odin.jwt.payload (lookup odin.jwt))
```
Comment (tleekkul) [inline tests/authnz.fg]: revert this change

File: views/app.register.fg [hunk 21]
```diff
@@ -101,81 +110,94 @@ POST odin/app/register / {
     "email": "fred@bowling.com"
 } 200
 
-# TODO: Register second time should return error
+# # TODO: Register second time should return error
```
Comment (tleekkul) [inline views/app.register.fg]: uncomment test and check it
